examples/computer_vision/shiftvit_pytorch/model_def.py

```diff
@@ -0,0 +1,117 @@
+"""
+ShiftViT code and paper:s
+https://github.com/microsoft/SPACH/blob/main/models/shiftvit.py
+https://arxiv.org/abs/2201.10801
+"""
+import logging
+import math
+import time
+import warnings
+
+import attrdict
+from determined.pytorch import (
+    DataLoader,
+    LRScheduler,
+    PyTorchCallback,
+    PyTorchTrial,
+    PyTorchTrialContext,
+)
+from timm.optim import create_optimizer
+from timm.scheduler import create_scheduler
+import torch
+from typing import Any, Dict, List, Optional, Sequence, Tuple, Union, cast
+
+import data
+import shiftvit  # shiftfit.py cloned via startup-hook.sh during container initialization.
+
+
+TorchData = Union[Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor]
+
+
+class ShiftViTTrial(PyTorchTrial):
+    def __init__(self, context: PyTorchTrialContext) -> None:
+        self.context = context
+
+        self.hparams = attrdict.AttrDict(self.context.get_hparams())
+        self.model_config = self.hparams.model
+        self.optimizer_config = self.hparams.optimizer
+        self.scheduler_config = self.hparams.scheduler
+        self.transform_config = self.hparams.transform
+        self.data_config = attrdict.AttrDict(self.context.get_data_config())
+        self.dataset_metadata = data.DATASET_METADATA_BY_NAME[self.data_config.dataset_name]
+
+        self.model = self.context.wrap_model(
+            shiftvit.ShiftViT(
+                **self.dataset_metadata.to_dict(),
+                **self.model_config,
+            )
+        )
+
+        # Use timm's create_xxx factories for the optimizer and scheduler.
+        optimizer = create_optimizer(self.optimizer_config, self.model)
+        self.optimizer = self.context.wrap_optimizer(optimizer)
+
+        # timm's scheduler expects to be stepped at the end of each epoch and to be passed the
+        # epoch_idx as an arg. Passing epoch_idx requires using MANUAL_STEP mode.
+        scheduler, _ = create_scheduler(self.scheduler_config, self.optimizer)
+        self.scheduler = self.context.wrap_lr_scheduler(
+            scheduler, step_mode=LRScheduler.StepMode.MANUAL_STEP
```

Comment:
is there a reason we are using manual step instead of step every epoch for the lr_scheduler?

Reply:
I'm using `timm`'s LR scheduler (as the ShiftVit paper does) which requires you to pass in the `epoch_idx` as an argument to `.step()`, and we only support such argument passing in `MANUAL_STEP` mode.

This is also why I was waiting to merge in the `epoch_idx` sync PR first; previously the `epoch_idx`'s became out of sync across workers.